src/v2/providers/storage.ts

```diff
@@ -0,0 +1,419 @@
+  if (arguments.length === 1) {
+    handler = optsOrHandler as (
+      event: CloudEvent<StorageObjectData>
+    ) => any | Promise<any>;
+    _onOperation(handler, metadataUpdatedEvent);
+  }
+  return _onOperation(
+    handler,
+    metadataUpdatedEvent,
+    optsOrHandler as string | StorageOptions
+  );
+}
+
+/** @hidden */
+export function _onOperation(
+  handler: (event: CloudEvent<StorageObjectData>) => any | Promise<any>,
+  eventType: string,
+  bucketOrOpts?: string | StorageOptions
+): CloudFunction<StorageObjectData> {
+  const [opts, bucket] = _getOptsAndBucket(bucketOrOpts);
+
+  const func = (raw: CloudEvent<unknown>) => {
+    return handler(raw as CloudEvent<StorageObjectData>);
+  };
+
+  func.run = handler;
+
+  // TypeScript doesn't recongize defineProperty as adding a property and complains
+  // that __trigger doesn't exist. We can either cast to any and lose all type safety
+  // or we can just assign a meaningless value before calling defineProperty.
+  func.__trigger = 'silence the transpiler';
+
+  Object.defineProperty(func, '__trigger', {
+    get: () => {
+      const baseOpts = options.optionsToTriggerAnnotations(
+        options.getGlobalOptions()
+      );
+      const specificOpts = options.optionsToTriggerAnnotations(opts);
+
+      return {
+        // TODO(colerogers): Remove "apiVersion" once the CLI has migrated to
+        // "platform"
+        apiVersion: 2,
+        platform: 'gcfv2',
+        ...baseOpts,
+        ...specificOpts,
+        labels: {
+          ...baseOpts?.labels,
+          ...specificOpts?.labels,
+        },
+        eventTrigger: {
+          eventType: provider + '.' + eventType,
+          resource: bucket, // TODO(colerogers): replace with bucket: 'my-bucket' when container contract is finished
+          service,
+        },
+      };
+    },
+  });
+
+  return func;
+}
+
+/** @hidden */
+export function _getOptsAndBucket(
+  bucketOrOpts?: string | StorageOptions
+): [options.EventHandlerOptions, string] {
+  let bucket: string;
+  let opts: options.EventHandlerOptions;
+  if (!bucketOrOpts) {
+    bucket = firebaseConfig().storageBucket;
+    opts = {};
+  } else if (typeof bucketOrOpts === 'string') {
+    bucket = bucketOrOpts;
+    opts = {};
+  } else {
+    bucket = bucketOrOpts.bucket || firebaseConfig().storageBucket;
+    opts = { ...bucketOrOpts };
+    delete (opts as any).bucket;
+  }
+
+  if (!bucket) {
+    throw new Error(
+      'Missing bucket name. If you are unit testing, please provide a bucket name' +
+        ' through set process.env.FIREBASE_CONFIG.'
+    );
+  }
+  if (!/^[a-z\d][a-z\d\\._-]{1,230}[a-z\d]$/.test(bucket)) {
+    throw new Error(`Invalid bucket name ${bucket}`);
+  }
+
+  return [opts, bucket];
+}
+
+/**
+ * An object within Google Cloud Storage.
+ * Ref: https://github.com/googleapis/google-cloudevents-nodejs/blob/main/cloud/storage/v1/StorageObjectData.ts
+ */
+export interface StorageObjectData {
```

Comment:
nit* Can this be declared at the top of the file instead?

Reply:
Moved to the top